src/kdf.cpp

```diff
@@ -0,0 +1,142 @@
+                                   const std::vector<uint8_t> &salt)
+    {
+        std::vector<uint8_t> derivedKey(outputLength);
+        const EVP_MD* digestFn = _getMDPtrFromDigestType(_hashFunction);
+        openssl::_PKCS5_PBKDF2_HMAC(reinterpret_cast<const char*>(password.data()), password.size(),
+                                    salt.data(), salt.size(),
+                                    _iterations, digestFn, outputLength, derivedKey.data());
+
+        return derivedKey;
+    }
+
+private:
+    openssl::DigestTypes _hashFunction;
+    uint32_t _iterations;
+};
+
+PBKDF2::PBKDF2(openssl::DigestTypes hashFunction, uint32_t iterations)
+{
+    _impl = std::make_unique<PBKDF2::Impl>(hashFunction, iterations);
+}
+
+PBKDF2::~PBKDF2() = default;
+
+std::vector<uint8_t> PBKDF2::deriveKey(const std::vector<uint8_t> &password,
+                                       const size_t outputLength,
+                                       const std::vector<uint8_t> &salt)
+{
+    return _impl->deriveKey(password, outputLength, salt);
+}
+
+PBKDF2::PBKDF2(PBKDF2&& other) = default;
+
+PBKDF2& PBKDF2::operator=(PBKDF2&& other) {
+    this->_impl = std::move(other._impl);
+    return *this;
+}
+
+PBKDF2::PBKDF2(const PBKDF2& other)
+    : _impl(std::make_unique<PBKDF2::Impl>(*(other._impl))) {}
+
+PBKDF2& PBKDF2::operator=(const PBKDF2& other) {
+    this->_impl = std::make_unique<PBKDF2::Impl>(*(other._impl));
+    return *this;
+}
+
+
+
+
+class X963KDF::Impl
+{
+public:
+    Impl(openssl::DigestTypes hashFunction) : _hashFunction(hashFunction) {}
+
+    ~Impl() = default;
+
+    std::vector<uint8_t> deriveKey(const std::vector<uint8_t> &password,
+                                   const size_t outputLength,
+                                   const std::vector<uint8_t> &salt)
+    {
+        std::vector<uint8_t> derivedKey(outputLength);
+        const EVP_MD* digestFn = _getMDPtrFromDigestType(_hashFunction);
+        openssl::_ECDH_KDF_X9_63(derivedKey.data(), outputLength, password.data(), password.size(), salt.data(),
+                                 salt.size(), digestFn);
+        return derivedKey;
+    }
+
+private:
+    openssl::DigestTypes _hashFunction;
+};
+
+X963KDF::X963KDF(openssl::DigestTypes hashFunction)
+{
+    _impl = std::make_unique<X963KDF::Impl>(hashFunction);
+}
+
+X963KDF::~X963KDF() = default;
+
+std::vector<uint8_t> X963KDF::deriveKey(const std::vector<uint8_t> &password, const size_t outputLength,
+                                        const std::vector<uint8_t> &salt)
+{
+    return _impl->deriveKey(password, outputLength, salt);
+}
+
+X963KDF::X963KDF(X963KDF&& other) = default;
+
+X963KDF::X963KDF(const X963KDF &other)
+    : _impl(std::make_unique<X963KDF::Impl>(*(other._impl))) {}
+
+X963KDF& X963KDF::operator=(X963KDF&& other) {
+    this->_impl = std::move(other._impl);
+    return *this;
+}
+
+X963KDF &X963KDF::operator=(const X963KDF &other)
```

Comment:
Style: & close to type and not to function name

Reply:
Done